gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java

```diff
@@ -924,6 +935,212 @@ public void deleteObjects(List<StorageResourceId> fullObjectNames) throws IOExce
+      for (FolderInfo currentFolder : folders) {
+        if (countOfChildren.get(currentFolder.getFolderName()) == 0L) {
+          addFolderResourceInBlockingQueue(
+              currentFolder, folderDeleteBlockingQueue, innerExceptions);
+        }
+      }
+
+      while (!countOfChildren.isEmpty()) {
+        while (!folderDeleteBlockingQueue.isEmpty()) {
+          FolderInfo folderToDelete = folderDeleteBlockingQueue.poll();
+          queueSingleFolderDelete(
+              folderToDelete,
+              countOfChildren,
+              folderDeleteBlockingQueue,
+              innerExceptions,
+              executor,
+              1);
+        }
+      }
+      // deleting any remaining resources
+      executor.shutdown();
+    }
+
+    if (!innerExceptions.isEmpty()) {
+      GoogleCloudStorageEventBus.postOnException();
+      throw GoogleCloudStorageExceptions.createCompositeException(innerExceptions);
+    }
+  }
+
+  /* Callable class specifically for deletion of folder resource */
+  public class DeleteFolder implements Callable<Void> {
+    private DeleteFolderRequest deleteFolderRequest;
+
+    @Override
+    public Void call() throws IOException {
+      storageControlClient.deleteFolder(deleteFolderRequest);
+      return null;
+    }
+
+    DeleteFolder(FolderInfo folder) {
+      this.deleteFolderRequest =
+          DeleteFolderRequest.newBuilder().setName(folder.toString()).build();
+    }
+  }
+
+  /** Helper function to delete a single folder resource */
+  public void queueSingleFolderDelete(
+      final FolderInfo folder,
+      HashMap<String, Long> countOfChildren,
+      BlockingQueue<FolderInfo> folderDeleteBlockingQueue,
+      final KeySetView<IOException, Boolean> innerExceptions,
+      final BatchExecutor batchExecutor,
+      final int attempt)
+      throws IOException {
+    final String bucketName = folder.getBucket();
+    final String folderName = folder.getFolderName();
+
+    checkArgument(
+        !Strings.isNullOrEmpty(bucketName),
+        String.format("No bucket for folder resource %s", folder.toString()));
+    checkArgument(
+        !Strings.isNullOrEmpty(folderName),
+        String.format("No folder path for folder resource %s", folder.toString()));
+
+    batchExecutor.queue(
+        new DeleteFolder(folder),
+        getDeletionCallback(
+            folder,
+            countOfChildren,
+            folderDeleteBlockingQueue,
+            innerExceptions,
+            batchExecutor,
+            attempt));
+  }
+
+  /**
+   * Helper function to add folderResource to blocking queue
+   *
+   * @param folderResource
+   * @param folderDeleteBlockingQueue blocking queue
+   */
+  private void addFolderResourceInBlockingQueue(
+      FolderInfo folderResource,
+      BlockingQueue<FolderInfo> folderDeleteBlockingQueue,
+      KeySetView<IOException, Boolean> innerExceptions) {
+    try {
+      // waits for a maximum of 1 second if the queue is overflowing
```

Comment:
Why would it overflow. What will happen to the item if it overflows?

Reply:
Since I am assuming that all folder resources are in memory, ideally I do not expect overflow to ever happen. But if overflow does happen, the insertion will wait for 1 second for atleast one element space in the queue. Since deletion is happening concurrently, I expect that some resource will be dequeued from the queue and thus create space for insertion.